contracts/test/oracles/default/BalancerLpStablePoolPriceOracleTest.t.sol

```diff
@@ -42,80 +44,123 @@ contract BalancerLpStablePoolPriceOracleTest is BaseTest {
 
   function afterForkSetUp() internal override {
     mpo = MasterPriceOracle(ap.getAddress("MasterPriceOracle"));
-    address wtoken = ap.getAddress("wtoken");
 
-    address[][] memory stablePoolUnderlyings = new address[][](3);
-    stablePoolUnderlyings[0] = asArray(stMATIC, wtoken);
-    stablePoolUnderlyings[1] = asArray(jBRL, BRZ);
-    stablePoolUnderlyings[2] = asArray(jEUR, agEUR);
-
-    address[] memory stableLps = asArray(stMATIC_WMATIC_pool, jBRL_BRZ_pool, jEUR_agEUR_pool);
     address[] memory linearLps = asArray(linearAaveUsdtPool, linearAaveUsdcPool, linearAaveDaiPool);
 
     stableLpOracle = new BalancerLpStablePoolPriceOracle();
-    stableLpOracle.initialize(stableLps, stablePoolUnderlyings);
+    stableLpOracle.initialize();
 
     linearLpOracle = new BalancerLpLinearPoolPriceOracle();
     linearLpOracle.initialize(linearLps);
-  }
-
-  function getLpTokenPrice(address lpToken, IPriceOracle oracle) internal returns (uint256) {
-    IPriceOracle[] memory oracles = new IPriceOracle[](1);
-    oracles[0] = oracle;
 
+    IPriceOracle[] memory lpOracles = new IPriceOracle[](linearLps.length);
+    for (uint256 i = 0; i < linearLps.length; i++) {
+      lpOracles[i] = linearLpOracle;
+    }
     vm.prank(mpo.admin());
-    mpo.add(asArray(lpToken), oracles);
-    emit log("added the oracle");
-    return mpo.price(lpToken);
+    mpo.add(linearLps, lpOracles);
   }
 
   function testReentrancyWmaticStmaticLpTokenOraclePrice() public fork(POLYGON_MAINNET) {
+    address vault = address(IBalancerStablePool(stMATIC_WMATIC_pool).getVault());
+    BalancerReentrancyAttacker reentrancyAttacker = new BalancerReentrancyAttacker(
+      IBalancerVault(vault),
+      mpo,
+      stMATIC_WMATIC_pool
+    );
+
     // add the oracle to the mpo for that LP token
-    {
-      IPriceOracle[] memory oracles = new IPriceOracle[](1);
-      oracles[0] = IPriceOracle(stableLpOracle);
+    IPriceOracle[] memory oracles = new IPriceOracle[](1);
+    oracles[0] = IPriceOracle(stableLpOracle);
+    vm.prank(mpo.admin());
+    mpo.add(asArray(stMATIC_WMATIC_pool), oracles);
 
-      vm.prank(mpo.admin());
-      mpo.add(asArray(stMATIC_WMATIC_pool), oracles);
-    }
+    // gives ETH to attacker
+    vm.deal(address(reentrancyAttacker), 5 ether);
+
+    // makes sure the address calling the attack is from attacker
+    vm.prank(address(reentrancyAttacker));
 
-    address vault = address(IBalancerStablePool(stMATIC_WMATIC_pool).getVault());
-    // raise the reentrancy flag for that vault
-    vm.store(vault, bytes32(uint256(0)), bytes32(uint256(2)));
     // should revert with the specific message
-    vm.expectRevert(bytes("Balancer vault view reentrancy"));
-    mpo.price(stMATIC_WMATIC_pool);
+    vm.expectRevert(bytes("BAL#420"));
```

Comment:
what does this error verify? the error 420 is this
`    uint256 internal constant ADDRESS_CANNOT_SEND_VALUE = 420;`

Reply:
apparently it is just the way it consumes any error thrown on the callback to return the leftovers

```
    ├─ [75010] BalancerReentrancyAttacker::startAttack() 
    │   ├─ [64239] 0xBA12222222228d8Ba445958a75a0704d566BF2C8::manageUserBalance{value: 2000000000000000000}([(0, 0x0000000000000000000000000000000000000000, 1000000000000000000, 0xa0Cb889707d426A7A386870A03bc70d1b0697598, 0xa0Cb889707d426A7A386870A03bc70d1b0697598)]) 
    │   │   ├─ emit InternalBalanceChanged(param0: BalancerReentrancyAttacker: [0xa0Cb889707d426A7A386870A03bc70d1b0697598], param1: 0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270, param2: 1000000000000000000)
    │   │   ├─ [6877] 0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270::deposit{value: 1000000000000000000}() 
    │   │   │   ├─ emit Deposit(from: 0xBA12222222228d8Ba445958a75a0704d566BF2C8, amount: 1000000000000000000)
    │   │   │   └─ ← ()
    │   │   ├─ [7841] BalancerReentrancyAttacker::receive{value: 1000000000000000000}() 
    │   │   │   ├─ [7302] 0xb9e1c2B011f252B9931BBA7fcee418b95b6Bdc31::price(0x8159462d255C1D24915CB51ec361F700174cD994) [staticcall]
    │   │   │   │   ├─ [6540] 0xcC6Aa628516BB46391B05B16E5058c877461cc76::price(0x8159462d255C1D24915CB51ec361F700174cD994) [delegatecall]
    │   │   │   │   │   ├─ [3571] BalancerLpStablePoolPriceOracle::price(0x8159462d255C1D24915CB51ec361F700174cD994) [staticcall]
    │   │   │   │   │   │   ├─ [341] 0x8159462d255C1D24915CB51ec361F700174cD994::getVault() [staticcall]
    │   │   │   │   │   │   │   └─ ← 0x000000000000000000000000ba12222222228d8ba445958a75a0704d566bf2c8
    │   │   │   │   │   │   ├─ [1015] 0xBA12222222228d8Ba445958a75a0704d566BF2C8::manageUserBalance([]) [staticcall]
    │   │   │   │   │   │   │   └─ ← "BAL#400"
    │   │   │   │   │   │   └─ ← "Balancer vault view reentrancy"
    │   │   │   │   │   └─ ← "Balancer vault view reentrancy"
    │   │   │   │   └─ ← "Balancer vault view reentrancy"
    │   │   │   └─ ← "Balancer vault view reentrancy"
    │   │   └─ ← "BAL#420"
    │   └─ ← "BAL#420"
```

Reply:
Just to leave our conversation registered, 420 is the correct error to expect in this case since the final error of the transaction is thrown by the user balance deposit function, and not by the mpo.price function.

Added the old reentrancy test so we can test the error message, and left the new reentrancy test so we can test the receive hook and simulate a real attack.